src/main/java/org/openrewrite/java/apache/httpclient5/AddTimeUnitArgument.java

```diff
@@ -0,0 +1,98 @@
+import org.openrewrite.java.MethodMatcher;
+import org.openrewrite.java.tree.J;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+
+@Value
+@EqualsAndHashCode(callSuper = true)
+public class AddTimeUnitArgument extends Recipe {
+
+    @Option(displayName = "Method pattern",
+            description = "A method pattern that is used to find matching method invocations.",
+            example = "org.apache.http.client.config.RequestConfig.Builder setConnectionRequestTimeout(int)")
+    String methodPattern;
+
+    @Option(displayName = "Time Unit",
+            description = "The TimeUnit enum value we want to add to the method invocation. Defaults to `MILLISECONDS`.",
+            example = "MILLISECONDS",
+            required = false)
+    @Nullable
+    TimeUnit timeUnit;
+
+    @Override
+    public String getDisplayName() {
+        return "Adds a TimeUnit argument to the matched method invocations";
+    }
+
+    @Override
+    public String getDescription() {
+        return "In Apache Http Client 5.x migration, an extra TimeUnit argument is required in the timeout and duration methods. " +
+                "Previously in 4.x, all these methods were implicitly having the timeout or duration expressed in milliseconds, " +
+                "but in 5.x the unit of the timeout or duration is required. So, by default this recipe adds " +
+                "`TimeUnit.MILLISECONDS`, it is possible to specify this as a parameter. Since all affected methods of " +
+                "the Apache Http Client 5.x migration only have one integer/long argument, the recipe applies with matched method " +
+                "invocations of exactly one parameter.";
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
+            final MethodMatcher matcher = new MethodMatcher(methodPattern);
+
+            @Override
+            public J.MethodInvocation visitMethodInvocation(J.MethodInvocation method, ExecutionContext executionContext) {
+                J.MethodInvocation m = super.visitMethodInvocation(method, executionContext);
+                if (matcher.matches(m)) {
+                    JavaTemplate template = JavaTemplate
+                            .builder(StringUtils.repeat("#{any()}, ", m.getArguments().size()) + "TimeUnit.#{}")
+                            .contextSensitive()
+                            .javaParser(JavaParser.fromJavaVersion().classpath("httpclient5", "httpcore5"))
+                            .imports("java.util.concurrent.TimeUnit")
+                            .build();
+
+                    List<Object> arguments = new ArrayList<>(m.getArguments());
+                    arguments.add(timeUnit != null ? timeUnit : TimeUnit.MILLISECONDS);
+
+                    m = template.apply(
+                            updateCursor(m),
+                            m.getCoordinates().replaceArguments(),
```

Comment:
I'm wondering what happens to the method type when we don't replace the method, but only add an argument; wouldn't the method type post-recipe still refer to the original (usually single argument) method instead of the new multiple argument method? I'd always thought that `replaceArguments` was only when you replace with the same number and type or arguments; not when you switch to a different method.

If the method type is updated correctly there's no need for a change, otherwise it might be best to replace the whole method rather than just the arguments.